### PR TITLE
[SYCL][NVPTX] Enable approximate div/sqrt with -ffast-math

### PR DESCRIPTION
**Pull request by frasercrmck**
The generation of approximate div/sqrt in the NVPTX backend is driven by the "unsafe-fp-math" function attribute. Presumably when the optimization was first added there was no way of getting at this information from ISel, or even that there was no suitable instruction-level representation to begin with.

Even today, the `afn` fast-math flag is appropriate for relaxing sqrt to an approximate version, but while some targets apply that reasoning to fdiv, it's not clear that's a valid reading of the language reference manual.

The problem with using the function attribute is that when inlining it must be set on *both* caller/callee functions, otherwise it is wiped.

Since CUDA's devicelib bytecode library has hundreds functions with unsafe-fp-math explicitly disabled, if we inline those functions into SYCL kernels, we disable the ability for the backend to generate approximate functions, not just inside the devicelib function but across the entire kernel.

This might explain why some performance reports we've received suggest that inlining certain maths functions can make things worse even when the CUDA compiler does the same thing (e.g., #14358 though this needs verified).

For this reason, presuambly, the NVPTX backend has two codegen options that override the function attribute and always generate approximate div/sqrt instructions. This patch thus explicitly sets these options when compiling SYCL for NVPTX GPUs. It does not do so for regular C/C++ or CUDA code to limit the wider impact on existing code.

### CODE REVIEW

**Comment by frasercrmck** [comment]
I note that the AMDGPU backend treats the fast-math `afn` flag as allowing approximate calculations on `fdiv`. The LangRef says:

``` quote
Allow substitution of approximate calculations for functions (sin, log, sqrt, etc). See floating-point intrinsic definitions for places where this can apply to LLVM’s intrinsic math functions.
```

so my reading was that it's not intended for `fdiv` as it's not "a function". It's not entirely clear either way so I can see AMD's reading too.

**Comment by frasercrmck** [comment]
Ping @intel/dpcpp-clang-driver-reviewers

**Review by mdtoguchi — APPROVED** [review]
The driver changes look OK to me.  Beyond just checking that the proper options are being passed to the device compilation, are there any codegen tests to verify the behavior?

**Comment by frasercrmck** [comment]
> The driver changes look OK to me. Beyond just checking that the proper options are being passed to the device compilation, are there any codegen tests to verify the behavior?

Thanks for the review!

Just to check - do you mean IR codegen or NVPTX codegen? There wouldn't be anything meaningful to check in LLVM IR because it doesn't affect the IR at all. The NVPTX codegen is tested elsewhere like [here (div)](https://github.com/intel/llvm/blob/sycl/llvm/test/CodeGen/NVPTX/div-ri.ll) and [here (sqrt)](https://github.com/intel/llvm/blob/sycl/llvm/test/CodeGen/NVPTX/sqrt-approx.ll).

So in short I think it's already covered?

**Comment by mdtoguchi** [comment]
> > The driver changes look OK to me. Beyond just checking that the proper options are being passed to the device compilation, are there any codegen tests to verify the behavior?
> 
> Thanks for the review!
> 
> Just to check - do you mean IR codegen or NVPTX codegen? There wouldn't be anything meaningful to check in LLVM IR because it doesn't affect the IR at all. The NVPTX codegen is tested elsewhere like [here (div)](https://github.com/intel/llvm/blob/sycl/llvm/test/CodeGen/NVPTX/div-ri.ll) and [here (sqrt)](https://github.com/intel/llvm/blob/sycl/llvm/test/CodeGen/NVPTX/sqrt-approx.ll).
> 
> So in short I think it's already covered?

Yes, NVPTX codegen.  Thanks!

**Comment by frasercrmck** [comment]
@intel/llvm-gatekeepers This is ready to merge, cheers
